### PR TITLE
KAFKA-3432; Cluster.update() thread-safety

### PR DESCRIPTION
**Pull request by ijuma**
Replace `update` with `withPartitions`, which returns a copy instead of mutating the instance.

### CODE REVIEW

**Comment by ijuma** [comment]
Review by @guozhangwang

File: clients/src/main/java/org/apache/kafka/common/Cluster.java [hunk 1]
```diff
@@ -118,39 +119,12 @@ public static Cluster bootstrap(List<InetSocketAddress> addresses) {
...
-                List<PartitionInfo> psNode = Utils.notNull(partsForNode.get(p.leader().id()));
-                psNode.add(p);
-            }
-        }
-
-        for (Map.Entry<Integer, List<PartitionInfo>> entry : partsForNode.entrySet())
-            this.partitionsByNode.put(entry.getKey(), Collections.unmodifiableList(entry.getValue()));
-
-        return this;
+    public Cluster withPartitions(Map<TopicPartition, PartitionInfo> partitions) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java:124]: Is this name clear enough or can we do better?

Comment (granthenke) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java:124]: I think the name is okay.

**Comment by guozhangwang** [comment]
LGTM. This is a much better approach for `StreamPartitionAssignor`.

**Comment by guozhangwang** [comment]
Merged to trunk and 0.10.0.

**Pull request closed without merging**